packages/babel-plugin-transform-destructuring/test/api.js

```diff
@@ -0,0 +1,71 @@
+import { parse } from "@babel/parser";
+import { transformFromAstSync } from "@babel/core";
+import _traverse from "@babel/traverse";
+const traverse = _traverse.default;
+
+import path from "path";
+import { fileURLToPath } from "url";
+const __dirname = path.dirname(fileURLToPath(import.meta.url));
+
+import glob from "glob";
+import { promises as fs } from "fs";
+
+const files = await new Promise((resolve, reject) => {
+  glob(
+    path.resolve(__dirname, "fixtures/**/@(exec|input).js"),
+    function (err, files) {
+      if (err) reject(err);
+      else resolve(files);
+    },
+  );
+});
+
+describe("regression-14438", function () {
+  describe.each(files)("fixture: %s", file => {
+    let src, originalAst, transformed;
+
+    beforeEach(async () => {
+      src = await fs.readFile(file, "utf-8");
+
+      originalAst = parse(src, {
+        allowImportExportEverywhere: true,
+        allowAwaitOutsideFunction: true,
+        allowReturnOutsideFunction: true,
+        allowSuperOutsideMethod: true,
+        allowUndeclaredExports: true,
+        errorRecovery: true,
+        attachComment: false,
+        createParenthesizedExpressions: true,
+      });
+
+      transformed = transformFromAstSync(originalAst, src, {
+        plugins: [
+          [
+            "@babel/plugin-transform-destructuring",
+            { useBuiltIns: true, loose: true },
+          ],
+        ],
+        configFile: false,
+        ast: true,
+        code: true,
+      });
+    });
+
+    it("has correct scope information after transform", function () {
+      if (transformed.ast) {
+        // exclude transformation failures: they're tested by the other cases
+        traverse.cache.clear();
+        traverse(transformed.ast, {
+          VariableDeclarator(path) {
+            const b = path.scope.getBinding(path.get("id").node.name);
```

Comment:
Further improvement idea: To make it a universal check, we should loop through `Object.keys(path.getBindingIdentifiers())`. The current check is ok for `transform-destructuring` since `path.get("id")` is always an identifier after transformed.

Reply:
Sure, but if the bindings don't exist, would that be returned in `path.getBindingIdentifiers()`?

Reply:
At the AST level, instead of LHS of the variable declarator we can also go for all identifiers that are present. But I couldn't think of a way to exclude globals or member expressions (e.g. `slice` in `arr.slice` or `Array` global) which definitely won't have scope bindings.

Reply:
`path.getBindingIdentifiers()` extracts ids from the AST, it does not access the scope info.

Reply:
> At the AST level, instead of LHS of the variable declarator we can also go for all identifiers that are present.

`hasBinding` will respect globals. `path.isReferencedIdentifier()` can exclude referenced identifiers. I think we can start from VariableDeclarator / Function params only.

The most strict check will be to compare the scope info after transform against a new scope created from scratch (without traverse cache), but I guess it will fail many, if not most current tests.